### DISCUSSION TITLE
Road to v1.4.0 - "Beta" release

### DISCUSSION THREAD

**Discussion by NiyaShy** [NiyaShy/XB1ControllerBatteryIndicator#55] [General]
Hey everyone,

to start off with the Discussions function of Github I have something special for you: an early pre-release/test version of what will become v1.4.0. I'm not uploading it to the normal release page because it's still in a rather unfinished state, most of the translations haven't been updated for the new features (so far only English and German are up to date) and the new stuff might need a bit more testing to make sure it always works as intended.

New stuff included:
- notification popup/overlay that can trigger on low battery and/or when you press the XB button (as long as you don't have the Win10 game bar configured to react to it!). See #52 for details and screenshots.
You now can also disable the toast notifications, might be interesting for folks who had issues with fullscreen programs/games.
- option to not show certain controller states on the systray icon - see #38 for details and screenshots.
- option to get notifications about new minor releases - I plan to use the 4th version number tier (for example 1.4.0.1) for versions that only got new languages added in the future. The "main" update check won't react to that.

As I said, this is still an early and unfinished release. You can try it out if the new features sound interesting to you, but otherwise I'd recommend to stick with 1.3.0 for now.

[XB1ControllerBatteryIndicator_1.4.0_beta_2024-07-15.zip](https://github.com/user-attachments/files/16238786/XB1ControllerBatteryIndicator_1.4.0_beta_2024-07-15.zip)

**Comment by blackhorse088** [comment]
Awesome work! Thanks!
Looks more official than Microsoft could ever make.
I'll be using it instead of the older version and report back any problems/suggestions.
So far so good 👍 

I disabled the Gamebar XB, but still can be used with Windows key + G
So we can still have both.

**Comment by blackhorse088** [comment]
Bug;
Popup doesn't trigger when XB is pressed, out of 10 presses it works 3-4 times randomly.

I'm not entirely sure if this related? But I just did firmware update for the controller 4.8.1923.0
I press the XB now but nothing happens most of the times and seems to work randomly, for once it worked 5/10 times and the other 3/10 times.
Maybe it was working perfectly before the firmware update?
Not sure, could be my luck, I didn't test it much to know.

We should see others feedback.

**Reply by NiyaShy** [reply]
Please be a bit more specific. Controller type/generation (XBOne or Series X)? Was the battery level already showing on the systray icon?
The latter is important, because I had to "disable" the popup for all states that don't have a battery level to report, otherwise it would default to empty.

I can only test with my 2 XBOne controllers (first gen without headset jack and 2nd/3rd(?) gen with discrete jack), and so far the popup triggered 100% of the time as long as battery level was already working.

**Comment by blackhorse088** [comment]
Yes, sorry.
Xbox one S controller. The one with audio jack (model 1708) bluetooth connection.
Battery level is showing fine on the tray icon as wireless controller.
 Just now it worked 10/10 then stopped for 4 presses and popped on the 5th. Works randomly.
 Tried reinstalling and re-pairing few times.
I think the firmware update messed it up :(

**Edit: Update:**
Friday-The13-rus said they changed the XB event trigger from "press" to "on release" to avoid spamming while holding the button.
But on mine, when it works, it triggers on press.
Hope this helps somehow?

**Reply by NiyaShy** [reply]
Hm... I just tried again with mine (model 1697, so a bit older than yours, connected via dongle), and it sometimes fails when I don't hit the XB button dead center. But it definitely only triggers on release, otherwise you would constantly get a trigger when you hold the button down.

**Reply by blackhorse088** [reply]
I think the firmware update changed the controller behavior.
Tried holding it down till controller turned off, didn't trigger, I had to spam it till it triggers (on press)

Would the app work on Xinput emulated PS5 controller?

A helpful link where you can find games that supports both Dx11 and 12:
https://www.pcgamingwiki.com/wiki/List_of_DirectX_12_games

**Reply by Friday-The13-rus** [reply]
Hi! I have 1708 model too firmware v 4.8.1923.0.
It is correct that popup displays on releasing button. And you should press a button a moment (about a half of second) in order to trigger popup.
I checked the controller via dongle and bluetooth connection. In case with dongle all is stable, popup triggers in all tries. In case with bluetooth the popup triggers in 80% tries, not sure why. Maybe connection via bluetooth is not so stable as via dongle.

**Reply by blackhorse088** [reply]
Mine on press and it works whenever it feels like.
Randomly, average of 40% 
But when it works, it work the moment I press the button.

**Reply by NiyaShy** [reply]
> Would the app work on Xinput emulated PS5 controller?

Most likely not since emulated controllers are usually disguised as wired versions, and those don't report a battery level. 
> A helpful link where you can find games that supports both Dx11 and 12:
> https://www.pcgamingwiki.com/wiki/List_of_DirectX_12_games

Thanks, wasn't even aware I own several games that support DX11 and 12 xD Will do some testing over the weekend.

**Comment by dreaNN1337** [comment]
Hey

I like the beta very much!
I found a bug: In Madden 21 (DirectX11) dont show the Overlay.
DirectX12 works without an problem.

Is there a fix insight?

Thanks for your whole work - I love it!

**Reply by NiyaShy** [reply]
Interesting... So the game runs fullscreen both times but the overlay only shows up when it runs in DX12 mode?
The thing is: I can't really reproduce/analyse the issue since I don't have any games available that support DX11 and DX12. So it's a bit difficult to find the cause...

**Reply by Friday-The13-rus** [reply]
Not so good news. I tested the popup on a few games with DX11 and DX12 and have not found any issues.
In general, the app does not interact with the game at all. It uses standard windows functions to display a window on the top of everything. If the game affect this functions (a game implementation can be different, even with dirty hacks) no chance to display the popup.

**Reply by NiyaShy** [reply]
Have to correct myself, thanks to the wiki link backhorse posted I found out that I actually own several games that support DX11 and 12.
And I can confirm your issue: in Borderlands 3 the overlay works fine in DX12 mode but won't appear in DX11 mode as long as BL3 is in focus.
Kinda weird that it only seems to affect some games since Fridays tests showed no issues.

**Reply by Friday-The13-rus** [reply]
Hmm... Interesting. Tried a few DX11 games that have installed. In Hades displayed. In TESO (window mode) displayed. In TESO (fullscreen) does not displayed.
Will try to figure out why.
Could you check if popup works on fullscreen/window mode too?

**Reply by NiyaShy** [reply]
Just cross checked with BL3. DX11 windowed and borderless window work fine, fullscreen not.

**Reply by Friday-The13-rus** [reply]
Made some research.
DX11 uses exclusive fullscreen mode. It means that while the game is focused nothing else cannot draw on the screen.
DX12 does not use exclusive mode, fullscreen is equal to borderless window mode.

If short - the popup will not be displayed for a lot of DX11 games in fullscreen mode. But will be for DX12 and windowed mode.
Some programs (like steam) injects code to a game in order to display overlay in exclusive fullscreen mode. Googled how injection can be implemented for our popup and at first glance it's a big amount of work.

Sorry, my fault. I didn't pay enough attention to the fullscreen mode when tested the popup, because I prefer the borderless window mode.

**Reply by blackhorse088** [reply]
If this any help, the Gamebar works on DX11 full screen.
But I noticed it creates another overlay window on top of game window which makes you lose the game focus.

**Comment by crackpotwolf** [comment]
Does not show the charge level on Series X+ Bluetooth...

**Reply by NiyaShy** [reply]
That seems to be a general issue with Series X controllers and Bluetooth and to my current knowledge I can't do/change anything about it. See #49.

**Comment by Simzy** [comment]
Ive been looking into seeing if i could make it as an overlay for a while now, i just came across this, thought it might be useful for this https://github.com/michel-pi/GameOverlay.Net

**Reply by NiyaShy** [reply]
Hm... do you mean as a permanent overlay? While it might be doable, I don't see much use in it because the battery level usually only changes after many hours. And for an overlay that only gets displayed for a short while we already have the solution Friday-The13-rus added.

**Reply by Simzy** [reply]
not as a permanent one, ive done one of them using rainmeter, i was thinking about the problem where directx 11 games have the forced fullscreen, maybe see if there is a way where the overlay can still be triggered by the button push but use what i linked as a way to get it to show up on the games

**Comment by ghost** [comment]
Hello

I made new translate in **Hungarian** language. Can you implement to new version?

Thanks,
Tuyeee

[Strings.hu.zip](https://github.com/NiyaShy/XB1ControllerBatteryIndicator/files/6447653/Strings.hu.zip)

**Reply by Tuyeee** [reply]
Its was me sorry

**Comment by NiyaShy** [comment]
Quick update: I've uploaded a new zip file (see first post) that contains a slightly updated binary. Not a lot changed, "just" some languages that were added/updated, a tiny code change that hopefully prevents rare crashes, and I also bumped the version number to 1.4.0 so you can keep it apart from the main 1.3 version 😅

**Comment by saulob** [comment]
Can you, please, add one "about" or any other place that shows the version number? I don't kno what version I'm using now :(

Maybe on the fist line one about saying the program name and version, there's no info about it. Show more info

![image](https://user-images.githubusercontent.com/3660977/233747644-8e5b86a2-fe29-4dcd-8862-980a22f10ef3.png)

Program name 
Version number
Github page (link)
Developer name :)


The basics ;)

"
XBox One Controller Battery Indicator
v1.4.0 beta 3
https://github.com/NiyaShy/XB1ControllerBatteryIndicator
@NiyaShy
"

**Reply by NiyaShy** [reply]
Something like this?
![grafik](https://user-images.githubusercontent.com/21028106/233802689-06bd9976-b8d7-48d1-ae12-e8a05812894a.png)
Getting just this simple menu to work was quite some work for me (ironically it's not that easy to get the program name and version from within the code itself...).
I'll think about adding a link to the website there, but IMHO it's not that important since you should know where you got the program from ;) Also, the update checker (if enabled) nudges you towards the website if there's something new available.
Dev name: definitely not. While I "own" the repository, so many others have contributed to the project that I'd never dare to claim all that for myself. 

Oh, and to answer the "question" which version you're running: judging by the entries in the context menu it's at least v1.3.0. For now you can find the exact version in the properties of the exe file itself.

**Comment by Arstidir** [comment]
Will it be possible to theme this a little bit more? Like Changing the shape, the Xbox icon etc. 
Im using Playnite FullScreen with a console like Ui and it would be awesome to integrate this overlay a little bit better.
I have something like this in mind:
![XB1ControllerBatteryIndicator](https://github.com/NiyaShy/XB1ControllerBatteryIndicator/assets/89221067/1ed1cd98-555a-49b9-9853-5de0aa2b3bcb)

**Reply by NiyaShy** [reply]
Possible: probably. Worth the effort: not that likely. Due to the code this project evolved from, the used graphics are all "baked" into the code and have to be in the ICO format to work. Allowing a custom icon would most likely require bigger changes so more graphics formats are supported and custom icons are the right size and don't go out of bounds.

**Comment by NiyaShy** [comment]
Just uploaded a new prerelease version (see first post). Most changes are in the background (like changing the used .NET version to 4.8 and updating a lot of Visual Studio packages). The most noticeable addition is the about menu that @saulob requested.

**Reply by saulob** [reply]
> Just uploaded a new prerelease version (see first post). Most changes are in the background (like changing the used .NET version to 4.8 and updating a lot of Visual Studio packages). The most noticeable addition is the about menu that @saulob requested.

Great work on the new prerelease version! The addition of the about menu is a fantastic improvement. This will help users easily identify which version of the program they're using, provide more details about where to download updates, and offer more information overall. Plus, it's a great marketing tool for you as the developer. Thanks for listening to the community's requests! 😊
